### PR TITLE
Support for late join after rekey/recovery

### PR DESCRIPTION
**Pull request by jumaffre**
Resolves https://github.com/microsoft/CCF/issues/315.

Until now, the leader was only sending the most recent ledger secret to the joiner during the join protocol. This was sufficient for late join at any time as long as we did not rekey or recover. 

This PR adds support for late join after rekey/recovery, whereby the whole history of ledger secrets is passed to the joiner in the join RPC response. Note that the ledger secrets no longer need to be sealed straight away on the joiner side but are sealed when the encryptor is compacted. The primary also passes a `public_only` boolean to the joining node that is used to setup Raft in `public_only` mode during recovery. 

Next: https://github.com/microsoft/CCF/issues/712

### CODE REVIEW

File: src/node/ledgersecrets.h [hunk 1]
```diff
@@ -43,68 +43,64 @@ namespace ccf
...
       // Generate fresh ledger encryption key
-      auto new_secret = std::make_unique<LedgerSecret>(true);
-      add_secret(1, std::move(new_secret), force_seal);
+      add_secret(1, LedgerSecret(true), force_seal);
     }
 
-    // Called when a node joins the network and get given the current ledger
-    // secrets
+    // Called when a node joins the network and get given the ledger secrets
+    // since the beggining of time
```
Comment (eddyashton) [inline src/node/ledgersecrets.h]: *beginning

File: src/node/ledgersecrets.h [hunk 2]
```diff
@@ -43,68 +43,64 @@ namespace ccf
     {}
   };
 
-  class LedgerSecrets
+  struct LedgerSecrets
```
Comment (eddyashton) [inline src/node/ledgersecrets.h]: Why is this changed from class to struct? We should add an access specifier for `secrets_map` and `operator==`, to make it clear whether they're intended to be `public` or `private`.

File: src/node/test/encryptor.cpp [hunk 3]
```diff
@@ -67,6 +69,7 @@ TEST_CASE("Additional data")
   // Setting 1 ledger secret, valid for version 1+
   uint64_t node_id = 0;
   auto secrets = std::make_shared<ccf::LedgerSecrets>();
+  secrets->set_secret(1, std::vector<uint8_t>(16, 0x42));
```
Comment (eddyashton) [inline src/node/test/encryptor.cpp]: If these secrets are unused in these unit tests, can we pass an empty vector or at least a named value (`dummy_secrets` or similar)?

File: tests/suite/test_suite.py [hunk 4]
```diff
@@ -8,20 +8,16 @@
 
 from inspect import signature, Parameter
 
-# TODO: For now, these are hardcoded. Indeed, late join after recovery is not yet supported.
-# https://github.com/microsoft/CCF/issues/315
+# TODO: For now, these are hardcoded.
+# https://github.com/microsoft/CCF/issues/712
```
Comment (eddyashton) [inline tests/suite/test_suite.py]: This hardcoded list is now missing most of the tests (`e2e_logging`, `add_node_untrusted_code`, `retire_node`). I think the hardcoded version should be doing 'everything' at least once?

**Comment by cimetrics** [comment]
late_join_rekey@[3996 aka 20200122.20](https://msrc-ccf.visualstudio.com/CCF/_build/results?buildId=3996&view=results) vs master ewma over 20 builds from [3780](https://msrc-ccf.visualstudio.com/CCF/_build/results?buildId=3780&view=results) to [3972](https://msrc-ccf.visualstudio.com/CCF/_build/results?buildId=3972&view=results)
images

File: src/node/ledgersecrets.h (hunk 2 above)
Comment (jumaffre) [reply src/node/ledgersecrets.h]: Good spot. I've reverted this to a `class`. The `secrets_map` is `public` so that we can serialised it in the Join RPC response.

File: src/node/test/encryptor.cpp (hunk 3 above)
Comment (jumaffre) [reply src/node/test/encryptor.cpp]: This is actually used as `secrets` is passed to the `encryptor` which uses these secrets for encryption/decryption.

File: tests/suite/test_suite.py (hunk 4 above)
Comment (jumaffre) [reply tests/suite/test_suite.py]: The test cases I have deleted are already tested as part of the `code_update_test` and `reconfiguration` end-to-end tests so I don't think they are needed here. Because the cost of setting up a network is fairly expensive, we might want to skip these standalone end-to-end tests at some point and only use the suite though.

For the `e2e_logging` test, my plan is to build a new transactions object that records all the transactions issued to the network since the beginning of the suite and can check whenever required (i.e. after recovery) that the transactions are still in CCF.
